changed dagger implementation to demo app as in dcl app

jaypatel9912 opened the pull request.

jayrcg committed:
changed dagger implementation as of dcl app

rkrsgithub commented on app/src/main/java/com/demoapp/adapter/CategoryDataAdapter.kt:
```diff
@@ -0,0 +1,76 @@
+import com.demoapp.model.DataEntity
+import de.hdodenhof.circleimageview.CircleImageView
+
+class CategoryDataAdapter(context: Context) :
```
Why we need context inside adapter ?

jaypatel9912 replied:
removed context from adapter

rkrsgithub commented on app/src/main/java/com/demoapp/database/DataRepository.kt:
```diff
@@ -0,0 +1,89 @@
+import retrofit2.Call
+import retrofit2.Callback
+
+class DataRepository(var dataDuo: DataDuo, val mService: RetrofitInterface) {
```
Where's injection ?

jaypatel9912 replied:
Injection added

rkrsgithub commented on app/src/main/java/com/demoapp/database/DataRepository.kt:
```diff
@@ -0,0 +1,89 @@
+
+class DataRepository(var dataDuo: DataDuo, val mService: RetrofitInterface) {
+
+    fun getLocalData(result: (List<DataEntity>) -> Unit) {
```
use suspend function instead of callback

jaypatel9912 replied:
added suspend function and removed callback

rkrsgithub commented on app/src/main/java/com/demoapp/model/CategoryEntity.kt:
```diff
@@ -0,0 +1,11 @@
+import androidx.room.PrimaryKey
+
+@Entity(tableName = "category_entity")
+class CategoryEntity(
```
```suggestion
data class CategoryEntity(
```

jaypatel9912 replied:
modified class to data class

rkrsgithub commented on app/src/main/java/com/demoapp/model/DataEntity.kt:
```diff
@@ -0,0 +1,17 @@
+import androidx.room.PrimaryKey
+
+@Entity(tableName = "data_entity")
+class DataEntity(
```
```suggestion
data class DataEntity(
```

jaypatel9912 replied:
modified class to data class

rkrsgithub commented on app/src/main/java/com/demoapp/viewmodel/MainActivityViewModel.kt:
```diff
@@ -0,0 +1,57 @@
+import com.demoapp.model.DataEntity
+import javax.inject.Inject
+
+class MainActivityViewModel(application: Application, dataDuo: DataDuo, mService: RetrofitInterface) : AndroidViewModel(application) {
```
add inject

rkrsgithub replied:
don't use AndroidViewModel

rkrsgithub commented on app/src/main/java/com/demoapp/viewmodel/MainActivityViewModel.kt:
```diff
@@ -0,0 +1,57 @@
+
+class MainActivityViewModel(application: Application, dataDuo: DataDuo, mService: RetrofitInterface) : AndroidViewModel(application) {
+
+    private var dataRepository = DataRepository(dataDuo, mService)
```
Do dagger injection

rkrsgithub replied:
This should be inject in viewmodel constructor

rkrsgithub commented on app/src/main/java/com/demoapp/viewmodel/MainActivityViewModel.kt:
```diff
@@ -0,0 +1,57 @@
+    private var dataRepository = DataRepository(dataDuo, mService)
+
+    private var allCategoryList: MutableLiveData<List<DataEntity>> = MutableLiveData()
+    fun getDataObserver(): MutableLiveData<List<DataEntity>> {
```
```suggestion
    fun getDataObserver(): LiveData<List<DataEntity>> {
```

jaypatel9912 replied:
added LiveData

rkrsgithub commented on app/src/main/java/com/demoapp/viewmodel/MainViewModelFactory.kt:
```diff
@@ -0,0 +1,21 @@
+import javax.inject.Inject
+
+@CommonScope
+class MainViewModelFactory @Inject constructor(val application: Application, val dataDuo: DataDuo, private val mService: RetrofitInterface) : ViewModelProvider.Factory {
+
+    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
+        if (modelClass.isAssignableFrom(MainActivityViewModel::class.java)) {
+            return MainActivityViewModel(application, dataDuo, mService) as T
+        }
+        throw IllegalArgumentException("Unknown ViewModel class")
+    }
+}
```
create common view model factory. check Payment method in DCL app

jaypatel9912 replied:
Created common view model factory

rkrsgithub commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="utf-8"?>
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Use Constraint layout

jaypatel9912 replied:
constraint layout added

rkrsgithub commented on app/src/main/res/layout/item_category.xml:
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8"?>
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Constraint layout

jaypatel9912 replied:
constraint layout added

rkrsgithub commented on app/src/main/res/layout/item_data.xml:
```diff
@@ -0,0 +1,54 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.appcompat.widget.LinearLayoutCompat xmlns:android="http://schemas.android.com/apk/res/android"
```
Use constraint layout

jaypatel9912 replied:
constraint layout added

rkrsgithub reviewed (changes requested):
Fix merge conflict

jayrcg committed:
Requested changes in PR are done

jaypatel9912 committed:
Merge branch 'master' into develop

rkrsgithub commented on app/src/main/java/com/demoapp/adapter/CategoryDataAdapter.kt:
```diff
@@ -11,6 +11,7 @@ import com.demoapp.R
 import com.demoapp.model.DataEntity
 import de.hdodenhof.circleimageview.CircleImageView
 
+
```
```suggestion
```

jaypatel9912 replied:
Removed space

rkrsgithub commented on app/src/main/java/com/demoapp/adapter/CategoryDataAdapter.kt:
```diff
@@ -24,6 +25,7 @@ class CategoryDataAdapter() :
         notifyDataSetChanged()
     }
 
+
```
```suggestion
```

jaypatel9912 replied:
Removed space

rkrsgithub commented on app/src/main/java/com/demoapp/database/DataRepository.kt:
```diff
@@ -1,5 +1,6 @@
 package com.demoapp.database
 
+
```
```suggestion
```

jaypatel9912 replied:
Removed space

rkrsgithub commented on app/src/main/java/com/demoapp/ui/MainActivity.kt:
```diff
@@ -23,6 +23,7 @@ class MainActivity : AppCompatActivity() {
         getViewModel(viewModelFactory)
     }
 
+
```
```suggestion
```

jaypatel9912 replied:
Removed space

rkrsgithub replied:
ok

rkrsgithub commented on app/src/main/java/com/demoapp/viewmodel/MainActivityViewModel.kt:
```diff
@@ -36,6 +36,7 @@ class MainActivityViewModel @Inject constructor(dataDuo: DataDuo, mService: Retr
     }
 
     init {
+
```
```suggestion
```

jaypatel9912 replied:
Removed space

rkrsgithub commented on app/src/main/java/com/demoapp/dagger/AppComponentProvider.kt:
```diff
@@ -0,0 +1,7 @@
+
+   fun  getAppComponent( ) : AppComponent
+
+}
```
Ensure enplane for all files

jaypatel9912 replied:
Raghu do you mean need to all activities or all the files requiring dependencies?

rkrsgithub replied:
nope. end line

jaypatel9912 replied:
removed end line

rkrsgithub replied:
Not to remove end line. we need enplane. Github should not show and icon like 🚫

rkrsgithub commented on app/src/main/java/com/demoapp/database/DataRepository.kt:
```diff
@@ -1,16 +1,17 @@
 
-    fun getLocalData(result : (List<DataEntity>) -> Unit) {
+    suspend fun getLocalData(): List<DataEntity> {
         var resultList = arrayListOf<DataEntity>()
```
Instead of additional list variable like this. use array map function to transform one data type to another

jaypatel9912 replied:
Removed extra variable and returned list with the use of arrayListOf().apply

rkrsgithub commented on app/src/main/java/com/demoapp/extension/ViewModelExtension.kt:
```diff
@@ -0,0 +1,42 @@
+/**
+ * Generic fragment extension function to be able to retrieve the viewModel given a [T] class.
+ */
+inline fun <reified T : ViewModel> Fragment.getViewModel(factory: ViewModelProvider.Factory, body: T.() -> Unit = {}): T {
+    val vm = ViewModelProviders.of(this, factory)[T::class.java]
+    vm.body()
+    return vm
+}
+
+
+inline fun <reified T : ViewModel> AppCompatActivity.getViewModel(factory: ViewModelProvider.Factory, body: T.() -> Unit = {}): T {
+    val vm = ViewModelProviders.of(this, factory)[T::class.java]
+    vm.body()
+    return vm
+}
```
We don't need this. it's already there in latest Kotlin version `viewModel()` and `activityViewModel()`

jaypatel9912 replied:
removed ViewModelExtension.kt class

rkrsgithub commented on app/src/main/java/com/demoapp/extension/ViewModelExtension.kt:
```diff
@@ -0,0 +1,42 @@
+/**
+ * Method to simplified how to set an Observer by just passing the [body] to be executed inside the observer.
+ */
+inline fun <T : Any, L : LiveData<T>> LifecycleOwner.observe(liveData: L, crossinline body: (T?) -> Unit) =
+    liveData.observe(this, Observer<T?> { t -> body(t) })
+
+fun <T : Any, L : LiveData<T>> LifecycleOwner.unObserve(liveData: L) = liveData.removeObservers(this)
+
+/**
+ * We need to restart the observer because it stays observing until the LifecycleOwner is destroyed.
+ * In the case of Fragments, they are not destroyed when the fragment is detached/reattached, and a new
+ * observer could be added every time the Fragment is shown and onActivityCreated() is executed.
+ * By doing this, we make sure we only have ONE observer at a time.
+ * Refer to: https://medium.com/@BladeCoder/architecture-components-pitfalls-part-1-9300dd969808
+ */
+inline fun <T : Any, L : LiveData<T>> LifecycleOwner.reObserve(liveData: L, crossinline body: (T?) -> Unit) {
+    unObserve(liveData)
+    observe(liveData, body)
+}
```
We are not going to use Live data. so this also not needed

jaypatel9912 replied:
removed ViewModelExtension.kt class

rkrsgithub commented on app/src/main/java/com/demoapp/ui/MainActivity.kt:
```diff
@@ -6,26 +6,36 @@ import android.view.MenuItem
+    @Inject
+    lateinit var viewModelFactory: ViewModelProvider.Factory
+
+    private val viewModel: MainActivityViewModel by lazy {
+        getViewModel(viewModelFactory)
+    }
```
```suggestion
    private val viewModel: MainActivityViewModel by viewModels { viewModelFactory }
```

jaypatel9912 replied:
Changed viewmodel initialization

rkrsgithub commented on app/src/main/java/com/demoapp/ui/MainActivity.kt:
```diff
@@ -6,26 +6,36 @@ import android.view.MenuItem
         supportActionBar?.setDisplayHomeAsUpEnabled(true)
 
-        viewModel = ViewModelProvider(this).get(MainActivityViewModel::class.java)
+        (application as App).getAppComponent().inject(this)
```
Do injection as the 1st line of code in onCreate

jaypatel9912 replied:
This change is done

rkrsgithub commented on app/src/main/java/com/demoapp/viewmodel/MainActivityViewModel.kt:
```diff
@@ -1,27 +1,23 @@
-
-    @Inject
-    lateinit var dataDuo: DataDuo
+class MainActivityViewModel @Inject constructor(dataDuo: DataDuo, mService: RetrofitInterface) :
```
We don't need direct access to Dao and Retrofit in viewModel. it should be in [Repository](https://developer.android.com/jetpack/guide) class.

jaypatel9912 replied:
Removed Dao and Retrofit from ViewModel and added to Repository class

rkrsgithub commented on app/src/main/java/com/demoapp/viewmodel/MainActivityViewModel.kt:
```diff
@@ -1,27 +1,23 @@
+    private var dataRepository = DataRepository(dataDuo, mService)
 
     private var allCategoryList: MutableLiveData<List<DataEntity>> = MutableLiveData()
-    fun getDataObserver(): MutableLiveData<List<DataEntity>> {
+    fun getDataObserver(): LiveData<List<DataEntity>> {
```
Use MutableStateFlow and StateFlow instead of LiveData

jaypatel9912 replied:
Added MutableStateFlow and StateFlow

jayrcg committed:
Pull request changes :
Flow instead of LiveData
removed unnecessary class
dependency inject in onCreate added
view model initialization changed
added dependency to DataRepository instead of view model

jayrcg committed:
Merge remote-tracking branch 'origin/develop' into develop

# Conflicts:
#	app/src/main/java/com/demoapp/adapter/CategoryDataAdapter.kt

rkrsgithub commented on app/src/main/java/com/demoapp/dagger/ViewModelModule.kt:
```diff
@@ -0,0 +1,23 @@
+    @Binds
+    abstract fun bindViewModelFactory(viewModelFactory: ViewModelFactory): ViewModelProvider.Factory
+
+}
```
end line

jaypatel9912 replied:
removed end line

rkrsgithub commented on app/src/main/java/com/demoapp/database/DataRepository.kt:
```diff
@@ -1,64 +1,54 @@
-                    category.name,
-                    ignoreCase = true
+        return arrayListOf<DataEntity>().apply {
+            categories.forEach {
```
No foreach allowed for data transformation

jaypatel9912 replied:
Flatmap used here instead of foreach and arraylist

jayrcg committed:
Removed unnecessary lines
data generated using flatmap and filter

rkrsgithub commented on app/src/main/java/com/demoapp/App.kt:
```diff
@@ -17,8 +18,7 @@ class App : Application() {
         return appComponent
     }
-
 }
```
end line. Change this in your Android Studio settings to ensure end line in all files.

rkrsgithub commented on app/src/main/java/com/demoapp/dagger/AppModule.kt:
```diff
@@ -29,5 +28,4 @@ class AppModule(private val application: Application) {
         return application.applicationContext
     }
-
 }
```
end line. Change this in your Android Studio settings to ensure end line in all files. 🚫

rkrsgithub commented on app/src/main/java/com/demoapp/database/DataDuo.kt:
```diff
@@ -6,24 +6,23 @@ import androidx.room.Query
     @Query("DELETE FROM data_entity")
-    fun removeData()
+    suspend fun removeData()
 }
```
end line. Change this in your Android Studio settings to ensure end line in all files.

rkrsgithub commented on app/src/main/java/com/demoapp/database/DataRepository.kt:
```diff
@@ -82,5 +68,4 @@ class DataRepository(val dataDuo: DataDuo, val mService: RetrofitInterface) {
         }
     }
-
 }
```
end line. Change this in your Android Studio settings to ensure end line in all files.

jayrcg committed:
enplane for all files

rkrsgithub commented on app/src/main/java/com/demoapp/model/CategoryEntity.kt:
```diff
@@ -5,7 +5,8 @@ import androidx.room.Entity
     @PrimaryKey(autoGenerate = true) @ColumnInfo(name = "id") val id: Int = 0,
     @ColumnInfo(name = "name") val name: String
 )
+
```
this seems more that one line at the end.

jaypatel9912 replied:
Extra line removed

rkrsgithub commented on app/src/main/java/com/demoapp/scope/CommonScope.kt:
```diff
@@ -0,0 +1,7 @@
+
+@Scope
+@kotlin.annotation.Retention(AnnotationRetention.RUNTIME)
+annotation class CommonScope
```
Why we need a custom scope ?

jaypatel9912 replied:
I have used it before but forgot to remove after
Removed it and pushed code

jayrcg committed:
Removed unused custom scope and removed extra line from CategoryEntity class

rkrsgithub reviewed (approved):
LG
